### PR TITLE
Merge all Prow CI scripts into one

### PR DESCRIPTION
**Pull request by travier**
This will increase code re-use and simplify future CI changes.

### CODE REVIEW

File: ci/prow-entrypoint.sh [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+set -xeuo pipefail
+
+# Main script acting as entrypoint for all Prow jobs
+
+build_test_qemu() {
+    # This script is called via build-test-qemu.sh which is the main Prow
+    # entrypoint for PRs to this repo, as well as for PRs on other repos,
+    # mainly coreos-assembler.  It assumes that `cosa init` has been run.
+
+    REDIRECTOR_URL="https://rhcos-redirector.apps.art.xq1c.p1.openshiftapps.com/art/storage/releases/"
```
Comment (miabbott) [inline ci/prow-entrypoint.sh:11]: Probably can make this a global for the whole script

File: ci/prow-entrypoint.sh [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    # we want to use RHEL 8.5 for testing until we can start using 8.6
+    # see https://github.com/openshift/release/pull/26193
+    curl -L http://base-"${ocpver_mut}"-rhel85.ocp.svc.cluster.local > src/config/ocp.repo
+    # fetch the 8.6 appstream repo to enable building of extensions
+    # see: https://github.com/openshift/os/issues/795
+    curl -Ls http://base-"${ocpver_mut}"-rhel86.ocp.svc.cluster.local | grep -A 3 rhel-8-appstream | sed '1,3 s/rhel-8-appstream/rhel-86-appstream/g' >> src/config/ocp.repo
+    cosa buildfetch --url=${prev_build_url}
+    cosa fetch
+    cosa build
+    cosa buildextend-extensions
```
Comment (miabbott) [inline ci/prow-entrypoint.sh:50]: Drain this into `build()`?

File: ci/prow-entrypoint.sh [hunk 3]
```diff
@@ -0,0 +1,182 @@
+#!/bin/bash
+set -xeuo pipefail
+
+# Main script acting as entrypoint for all Prow jobs
+
+# Global variables
+REDIRECTOR_URL="https://rhcos-redirector.apps.art.xq1c.p1.openshiftapps.com/art/storage/releases/"
+
+# This script is used to update the /etc/passwd file within the COSA container
```
Comment (miabbott) [inline ci/prow-entrypoint.sh:9]: nit

```suggestion
# This function is used to update the /etc/passwd file within the COSA container
```

File: ci/prow-entrypoint.sh [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    cosa kola run --parallel 2
+    # Build metal + installer now so we can test them
+    cosa buildextend-metal && cosa buildextend-metal4k && cosa buildextend-live
+    # compress the metal and metal4k images now so we're testing
+    # installs with the image format we ship
+    cosa compress --artifact=metal --artifact=metal4k
+    # Running testiso scenarios on metal artifact
+    # Skip the following scenarios: iso-install,iso-offline-install,iso-live-login,iso-as-disk
+    # See: https://github.com/openshift/os/issues/666
+    kola testiso -S --scenarios pxe-install,pxe-offline-install --output-dir tmp/kola-metal
```
[the remark this answers is not in the document]
Comment (miabbott) [reply ci/prow-entrypoint.sh:87]: Latest `cosa` shows:

```
      --scenarios strings   Test scenarios (also available: [iso-install]) (default [pxe-install,iso-offline-install,pxe-offline-install,iso-live-login,iso-as-disk,miniso-install,miniso-install-nm])
```

File: ci/prow-entrypoint.sh [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+    # Build metal + installer now so we can test them
+    cosa buildextend-metal && cosa buildextend-metal4k && cosa buildextend-live
+
+    # compress the metal and metal4k images now so we're testing
+    # installs with the image format we ship
+    cosa compress --artifact=metal --artifact=metal4k
+
+    # Running testiso scenarios on metal artifact
+    # Skip the following scenarios: iso-install,iso-offline-install,iso-live-login,iso-as-disk
+    # See: https://github.com/openshift/os/issues/666
```
Comment (miabbott) [inline ci/prow-entrypoint.sh:80]: Comment shouldn't be applicable anymore

File: ci/prow-entrypoint.sh [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+    kola testiso -S --scenarios pxe-install,pxe-offline-install,iso-install,iso-offline-install,iso-live-login,iso-as-disk --output-dir tmp/kola-metal
+
+    # iso-install scenario to sanity-check the metal4k media
+    # Skip all the testiso scenarios for metal4k + UEFI
+    # See: https://github.com/openshift/os/issues/666
```
Comment (miabbott) [inline ci/prow-entrypoint.sh:85]: Same as above

File: ci/prow-entrypoint.sh [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+    # Fetch the previous build
+    cosa buildfetch --url=${prev_build_url}
+
+    # Fetch the repos corresponding to the release we are building
+    rhelver=$(rpm-ostree compose tree --print-only manifest.yaml | jq -r '.["automatic-version-prefix"]' | cut -f2 -d.)
+    curl -L "http://base-${ocpver_mut}-rhel${rhelver}.ocp.svc.cluster.local" > "src/config/ocp.repo"
+
+    cosa fetch
+    cosa build
+    cosa buildextend-extensions
```
Comment (miabbott) [inline ci/prow-entrypoint.sh:55]: Maybe `cosa kola run --basic-qemu-scenarios` here?

File: ci/prow-entrypoint.sh [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+    cosa kola run basic
+}
+
+kola_test_basic_all() {
```
[the remark this answers is not in the document]
Comment (travier) [reply ci/prow-entrypoint.sh:59]: It run in parallel of other tests: metal, upgrade, etc.

File: ci/prow-entrypoint.sh (hunk 7 above)
Comment (travier) [reply ci/prow-entrypoint.sh:58]: I ended up undoing that as this function will run in two different steps in Prow (during the build phase and during the test phase) thus we can not do kola testing here without slowing down the build/test too much.

File: ci/prow-entrypoint.sh [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+    fi
+}
+
+# Basic syntaxt validation for manifests
+validate() {
+    # Create a temporary copy
+    workdir="$(mktemp -d)"
+    echo "Using $workdir as working directory"
+    cd "$workdir"
+    # Figure out if we are running from the COSA image or directly from the Prow src image
```
Comment (miabbott) [inline ci/prow-entrypoint.sh:108]: Nice! :+1:

**Comment by travier** [comment]
This is failing on:
```
+ curl -L http://base-4-11-rhel86.ocp.svc.cluster.local
/src/ci/prow-entrypoint.sh: line 53: src/config/ocp.repo: Permission denied
```

**Comment by miabbott** [comment]
I think the permissions problem is due to this:

```
+ cosa init --transient /src
info: Missing CAP_SYS_ADMIN; using virt
+ mkdir -p src
+ cd src
+ test -e config
+ case "${source}" in
+ ln -s /src/ config
```

The `/src` checkout must be created as `root` when the job starts, but we are running as `builder`, so we can't write to that location.

Maybe we hack around this by copying `/src` to `$(mktemp -d)/src` as a workaround and init from that new repo instead?

**Comment by miabbott** [comment]
Looks promising

```
[36mINFO[0m[2022-06-09T18:03:55Z] Building src                                 
[36mINFO[0m[2022-06-09T18:06:00Z] Build src succeeded after 2m1s               
[36mINFO[0m[2022-06-09T18:06:00Z] Building build-test-qemu-img                 
[36mINFO[0m[2022-06-09T18:11:40Z] Build build-test-qemu-img succeeded after 5m36s 
[36mINFO[0m[2022-06-09T18:11:40Z] Running multi-stage test test-qemu-kola      
[36mINFO[0m[2022-06-09T18:11:40Z] Running multi-stage phase pre                
[36mINFO[0m[2022-06-09T18:11:40Z] Step phase pre succeeded after 0s.           
[36mINFO[0m[2022-06-09T18:11:40Z] Running multi-stage phase test               
[36mINFO[0m[2022-06-09T18:11:40Z] Running step test-qemu-kola-test.            
```

**Comment by travier** [comment]
Next step in https://github.com/openshift/release/pull/29329 once this one is ready

**Comment by miabbott** [comment]
/lgtm

Great job, @travier !

**Pull request merged**